app/lib/discourse_automation/scriptable.rb

```diff
@@ -0,0 +1,116 @@
+module DiscourseAutomation
+  class Scriptable
+    attr_reader :fields, :automation
+
+    def initialize(automation)
+      @automation = automation
+      @placeholders = [:site_title]
+      @version = 0
+      @fields = []
+      @triggerables = []
+      @script = Proc.new {}
+
+      eval!
+    end
+
+    def eval!
+      public_send("__scriptable_#{automation.script.underscore}")
+      self
+    end
+
+    def name
+      @automation.script
+    end
+
+    def placeholders
+      @placeholders.uniq.compact
+    end
+
+    def placeholder(placeholder)
+      @placeholders << placeholder
+    end
+
+    def version(*args)
+      if args.present?
+        @version, = args
+      else
+        @version
+      end
+    end
+
+    def triggerables(*args)
+      if args.present?
+        @triggerables, = args
+      else
+        @triggerables
+      end
+    end
+
+    def script(&block)
+      if block_given?
+        @script = block
+      else
+        @script
+      end
+    end
+
+    def field(name, component:, accepts_placeholders: false)
+      @fields << {
+        name: name,
+        component: component,
+        accepts_placeholders: accepts_placeholders
+      }
+    end
+
+    def utils
+      Utils
+    end
+
+    module Utils
+      def self.apply_placeholders(input, map)
+        map[:site_title] = SiteSetting.title
+
+        map.each do |key, value|
+          input = input.gsub("%%#{key.upcase}%%", value)
```

Comment:
I _think_ a safer & faster way to do the replacements would be to `.dup` the `input` parameter and then do `.gsub!` to update the same string

```suggestion
          input.gsub!("%%#{key.upcase}%%", value)
```